sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -121,98 +121,233 @@ Use ``node make help`` to list all available commands.
+BokehJS-related pull requests that result in changes to the visual output generated by
+BokehJS should include visual baseline comparison tests.
+
+In the background, BokehJS' testing framework runs a headless browser and takes
+screenshots of the browser's output. The testing framework then compares the visual
+output to each test's dedicated baseline files.
+
+Each test in ``test:integration`` consists of two types of baseline comparisons:
+
+Textual baseline comparison
+  For each test, the testing framework compares the pixel location of certain elements
+  in the visual output to pixel locations in a baseline file. The pixel locations are a
+  set of bounding boxes. They are stored as plain text in each test's ``.blf`` files.
+
+Visual baseline comparison
+  For each test, the testing suite does a pixel-by-pixel comparison of a screenshot
+  and a baseline image. These baseline images are stored as ``.png`` files. In contrast
+  to textual baseline comparisons, visual baseline comparisons are platform-dependent.
+  Even minor differences in font rendering, for example, will make the pixel-by-pixel
+  comparison fail.
+
+The visual baseline comparison tests are located in the
+:bokeh-tree:`bokehjs/test/integration/` folder and sub-folders.
+:ref:`Bokeh's CI <devguide_testing_ci>` runs these tests on Linux, macOS, and Windows
+environments. The baseline files for each environment are located in the
+:bokeh-tree:`bokehjs/test/baselines/` folder.
+
+Follow these steps to write new visual tests or update existing tests:
+
+1. Create or update visual testing scripts:
+    To write a visual test for BokehJS' testing suite, start by importing the
+    ``display()`` and ``fig()`` functions from the testing framework's ``_util`` module
+    (located in :bokeh-tree:`bokehjs/test/integration/`):
+
+    .. code-block:: TypeScript
+
+        import {display, fig} from "./_util"
+
+    When writing tests, replace BokehJS' standard ``show()`` function with the
+    ``display()`` function in ``_util``. ``display()`` accepts the same arguments as
+    ``show()`` but also captures the visual output for comparison.
+
+    Similarly, replace BokehJS' standard ``figure()`` with the ``fig()`` function in
+    ``_util``. ``fig()`` expects an array of ``[width, height]`` as the first argument,
+    followed by the same arguments as ``figure()``. To keep visual tests as efficient as
+    possible, you should only use ``width`` and ``height``.
+
+    Keep the width and height of your testing plot as small as possible while still
+    being able to see the details you want to test with the naked eye. Try to keep the
+    number of elements on your plot to a minimum.
+
+    Follow this general pattern for visual tests:
+
+    .. code-block:: TypeScript
+
+      describe("Your Object", () => {
+        it("should show certain behavior", async () => {
+          const p = fig([width, height])
```

Comment:
`const p = fig([width, height], {figure_attrs})` (or something like that) would be better.